src/key-derivation/derived-key.ts

```diff
@@ -102,22 +106,22 @@ export class DerivedKeyOptions implements IDerivedKey {
     this.hash = options.hash || 'SHA256';
   }
 
-  public serialize(): string {
+  public serialize(serializtionVersion: SerializationFormat = SerializationFormat.latest_version): string {
     // keys taken from ruby lib
     return serializeDerivedKeyOptions(this.strategy, {
       iv: stringAsBinaryBuffer(this.salt), // ensures proper yaml serialization
       i: this.iterations,
       l: this.length,
       hash: this.hash
-    });
+    }, serializtionVersion);
   }
 
   public deriveKey(key: string): Promise<string> {
     const hash: string = this.hash.toLocaleLowerCase();
     const digest = md[hash as 'sha256'].create();
     return new Promise((resolve, reject) => {
       return pkcs5.pbkdf2(
-        key,
+        encodeUtf8(key),
```

Comment:
How come this is required now?

Reply:
**Summary:**  JavaScript default encoding is UTF-16 and other ports(Ruby, Elixir) uses UTF-8 because of that if we don't convert the passphrase to UTF-8, it produces derived key that is not compatible with other ports.

Two problems 

- encrypted data using a javascript port that includes a special character in passphrases e.g. ù it produces a result that can not be decrypted by (Ruby or Elixir) port and our **downloader service will fail**.
- encrypted data using the other ports that include a special character in passphrases e.g. ù  cant be decrypted in javascript port.

Example :

Generating derived key from passphrase without converting it to UTF-8 string :

JavaScript default encoding (UTF-16)
JavaScript uses an encoding that uses two bytes to store each character, which translates roughly to an encoding called UCS-2, or another one called UTF-16 .

Example

passphrase:
"Tiramisù Hans Zemlak"

The binary version of passphrase with UTF-16: 
Buffer.from(passphrase, 'binary')
Buffer(20) [84,105,114,97,109,105,115,**249**,32,72,97,110,115,32,90,101,109,108,97,107]

DerivedKey Produced from above passphrase:
Buffer.from(derivedKey, 'binary')
[81,242,224,121,157,16,201,30,68,130,108,53,173,174,60,1,171,206,124,115,71,219,147,82,77,50,26,251,163,202,169,33]


Generating derived key from passphrase with converting it to UTF-8 string :

encodeUtf8(passphrase)
"TiramisÃ¹ Hans Zemlak"

Buffer.from(encodeUtf8(passphrase), 'binary')
Buffer(21) [84,105,114,97,109,105,115,**195,185**,32,72,97,110,115,32,90,101,109,108,97,107]

DerivedKey Produced from above passphrase:
Buffer.from(derivedKey, 'binary')
Buffer(32)
[191,160,240,24,55,89,187,151,146,102,179,182,244,116,167,111,228,142,10,210,16,80,222,168,223,155,171,22,30,23,227,141]